Error page upgrades

DXCanas opened the pull request:
<!--
 1. Following guidance below, replace …'s with your own words
 2. After saving the PR, tick of completed checklist items
 3. Skip checklist items that are not applicable or not necessary
 4. Delete instruction/comment blocks
-->

### Summary
<!--
 * description of the change
 * manual verification steps performed
 * screenshots if the PR affects the UI
-->

New error page with associated screen shots:

![error page](https://user-images.githubusercontent.com/9877852/43555038-bfb5cca0-95ac-11e8-9c89-43234939098e.png)

![report error modal](https://user-images.githubusercontent.com/9877852/43554376-ce2a0c00-95a8-11e8-9fc5-32714b91c2d9.png)


Switched to large due to https://github.com/learningequality/kolibri/issues/3200#issuecomment-408276495

Here's what it looks like medium:
![error report screencap, medium](https://user-images.githubusercontent.com/9877852/43432239-9b48801a-9426-11e8-99b3-63fb4f82cd7a.png)


### Reviewer guidance
<!--
 * how can a reviewer test these changes?
 * are there any risky areas that deserve extra testing
-->

Tried inducing errors, but found that the `CORE_SET_ERROR` mutation couldn't be called? Anyone else seeing this? @jonboiser might it have something to do with vuex modules?

### References
<!--
 * references to related issues and PRs
 * links to mockups or specs for new features
 * links to the diffs for any dependency updates, e.g. in iceqube or the perseus plugin
-->

resolves #3200 

----

### Contributor Checklist

- [x] Contributor has fully tested the PR manually
- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [x] If there are any front-end changes, before/after screenshots are included
- [x] If this is an important user-facing change, PR or related issue has a 'changelog' label

### Reviewer Checklist

- [ ] Automated test coverage is satisfactory
- [ ] Reviewer has fully tested the PR manually
- [ ] PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- [ ] External dependencies files were updated (`yarn` and `pip`)
- [ ] Documentation is updated
- [ ] Link to diff of internal dependency change is included
- [ ] CHANGELOG.rst is updated for high-level changes
- [ ] Contributor is in AUTHORS.rst

DXCanas committed:
Having window title indicate Kolibri.

DXCanas committed:
Renaming error-box to app-error.

DXCanas committed:
Dichotomizing error page and body slot.

DXCanas committed:
Making "wrapper" class a little less generic.

DXCanas committed:
Switching padding to the body wrapper.

DXCanas committed:
Adding authMessage to app-error, annotating existing template.

DXCanas committed:
Boilerplating error page.

DXCanas committed:
Using props for loader position rather than hard-coded values.
Makes changing those values simpler.

benjaoming commented:
This sounds nice!

Can you provide a screen shot?

Why is it targeted for the `error-upgrades` branch?

christianmemije commented on kolibri/core/assets/src/views/app-body.vue:
```diff
@@ -10,9 +10,17 @@
+    >
+      <app-error v-if="error" />
+
+      <slot v-else>
```
Same line

DXCanas replied:
Can we wait till this PR goes into develop to get this granular?

indirectlylit commented:
@benjaoming see:

* https://docs.google.com/document/d/11K2_SrKfHFSnapAMcUDSu0Si3pD758gTm56UPw2BMc8/edit#heading=h.a15vlk954e6v
* https://drive.google.com/drive/folders/1Nk5B6MAYDND6rRYgCwOuZue5RSgf-emw

DXCanas commented:
@benjaoming This is WIP work for a feature branch that @ralphiee22 @christianmemije and I are working on :) Not ready for merge into develop. Beginning to think opening a PR may have been the wrong move.

benjaoming commented:
> Beginning to think opening a PR may have been the wrong move.

Ah no that's always fine if you're looking for inputs :) I was just asking because the final milestone was unclear and could guide the input that you might want -- which is why I was asking for a screenshot. Would seem like a good way to gather feedback.

But yeah, if you think *Reviewer guidance* cannot be filled in, then okay a PR may not be the way to go :)

DXCanas committed:
Single-lining <slot>

DXCanas committed:
Boilerplating error details modal

DXCanas committed:
Transitioning app-error vuefile to folder.

DXCanas committed:
Breaking modal out to its own component.

DXCanas committed:
Handling auth errors in core-base rather than app-body/app-error.

DXCanas committed:
Using core-base auth message in coach.

Signed-off-by: David Canas <david@learningequality.org>

DXCanas committed:
Using core-base auth message in facility_management

DXCanas committed:
Boilerplating automatic backend auth error display.

DXCanas committed:
Updating to match new design.
Error Details -> Report Error.
Moving copy options to bottom of modal.

DXCanas committed:
Enabling "download as text" link.
Restructuring page according to updated mockups.
Updating core-error-bg variable.

DXCanas committed:
Making forum link wrap.

DXCanas committed:
Having email link display correctly.

DXCanas committed:
Adding 'copy to clipboard' functionality.

DXCanas committed:
Switching from <p> to <div>.
Letting prettier do its thing.

DXCanas committed:
Forcing error file download on link click.
Rather than opening file in new window.
Adding optional prop to k-external-link.
Adding default file name.

DXCanas committed:
Clean up dummy error message.

DXCanas committed:
Converting clipboard compatibility check from data -> computed prop.

DXCanas committed:
Adding clipboard copy confirmation via snackbar.

DXCanas committed:
Updating strings to match new mockup

DXCanas committed:
Replacing Kolibri logo with error bird.

DXCanas committed:
Settling for a single view of the error reporting modal
(for now)

DXCanas committed:
Adding max width to k-modal.

DXCanas committed:
Correcting error code used to detect backend "not authorized" errors.

DXCanas commented:
Flagging: In an effort to induce an error, I noticed that called to the `CORE_SET_ERROR` mutation aren't actually being committed?

DXCanas committed:
Merge remote-tracking branch 'LE/error-upgrades' into error-page-upgrades
Dealing with file name changes.

# Conflicts:
#	kolibri/core/assets/src/views/AppBody.vue
#	kolibri/core/assets/src/views/CoreBase.vue
#	kolibri/core/assets/src/views/ErrorBox.vue
#	kolibri/core/assets/src/views/KModal.vue
#	kolibri/plugins/coach/assets/src/views/CoachIndex.vue
#	kolibri/plugins/facility_management/assets/src/views/FacilityIndex.vue

DXCanas committed:
Fixing "home" link.

DXCanas committed:
Having error display with new state.

DXCanas committed:
Pretty-printing error in report modal.

DXCanas committed:
Adding refresh button.
Removing unused components.
Removing unnecessary styles.

DXCanas committed:
Adding role to error page for accessibility.

DXCanas commented:
@khangmach Mind taking a quick look at those screencaps?

khangmach reviewed (approved):
very nice!!

DXCanas committed:
Stopping non-form errors from breaking js.

Signed-off-by: David Canas <david@learningequality.org>

DXCanas committed:
Fixing loginError test.

benjaoming commented:
@DXCanas This stuff looks great and is a great way of mixing detailed error reports with something that doesn't make users think that everything is wrong and they should move ten steps away from the computer :)

I know that this is perhaps a bit late for input, but the Refresh button:

1) Does it work if POST data was submitted? Can it loose other types of instance data being stored?
2) Does refreshing risk to loose the original error data?

DXCanas commented:
@benjaoming 

> Does it work if POST data was submitted? Can it loose other types of instance data being stored?

It's a cache-busting URL-based refresh. So it'll lose any data _not_ encoded in the URL that was made anytime before or after the URL is loaded on browser. 

I think this behavior is expected, though. There are instances of requests being constructed based off of information being cached in the frontend. If the cache is out of date, and a request is made based on that faulty information, it'll likely result in a server-side error. In this case, we would want to refresh the page, rebuilding the state and cache.

> Does refreshing risk to loose the original error data?

Yes, at least on the frontend. The data error data is held in the store in the frontend, so if we want a record of the error it would be a matter of logging the request/response in a log or something. 

I think it's also important to note that the changes made in the PR have no effect on the way the error is being stored - frontend or backend.
